### PR TITLE
feat: serialised standard extensions

### LINKED ISSUE #1228: All extensions should have a machine readable source of truth
We need to be able to exchange extension definitions between different hugr tools. 

All extension documentation should be generated by loading the extension and `.print_docs_markdown()`ing it.

Extensions are already serialisable, with the caveat of #1057 
YAML Declarative extensions are one source of machine readable extensions, but not the only one. Extensions should be exchanged in JSON and validated against a versioned schema, just like Hugrs.

Tasks:
 - [ ] #1057 
 - [x] Define a JSON schema for extensions
 - [ ] Remove all extension docs from the spec and replace with markdown generated by extension definitions.
 - [ ] Guppy is able to load an Extension from JSON and automagically provide the ops and types

### PR DESCRIPTION
**Pull request by ss2165**
Closes #1228 

Add cli for generating serialised extensions, and justfile/CI for checking checked in definitions match. 

Switched from `HashMap` to `BTreeMap` in `Extension` to ensure serialization happens in alphabetical order.

### CODE REVIEW

File: hugr-cli/src/extensions.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            path.set_extension("json");
+
+            std::fs::create_dir_all(path.clone().parent().unwrap()).unwrap();
+            // file buffer
+            let mut file = std::fs::File::create(&path).unwrap();
+
+            serde_json::to_writer_pretty(&mut file, &ext).unwrap();
+
+            // write newline, for pre-commit checks.
+            file.write_all(b"\n").unwrap();
```
Comment (mark-koch) [inline hugr-cli/src/extensions.rs:44]: Why is this related to the pre-commit check?

File: justfile [hunk 2]
```diff
@@ -55,6 +55,9 @@ update-schema:
     poetry update
     poetry run python scripts/generate_schema.py specification/schema/
 
+# Generate serialized declarations for the standard extensions and prelude.
+gen-extensions:
+    cargo build -p hugr-cli && poetry run ./target/debug/hugr gen-extensions -o specification/std_extensions
```
Comment (mark-koch) [inline justfile:60]: Why not `cargo run`?

**Review by mark-koch — APPROVED** [review]

File: hugr-cli/src/extensions.rs (hunk 1 above)
Comment (ss2165) [reply hugr-cli/src/extensions.rs:44]: expanded comment
